### PR TITLE
feature/CPLP- 753 decline application request

### PR DESCRIPTION
**Pull request by qxz2mqe**
https://jira.catena-x.net/browse/CPLP-753

### CODE REVIEW

**Comment by evegufy** [comment]
@Phil91 Could you please review? Thank you!

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalBackendDBAccess.cs [hunk 1]
```diff
@@ -482,6 +482,8 @@ public async IAsyncEnumerable<Guid> GetUserRoleIdsUntrackedAsync(IDictionary<str
                 .Select(app => app.AppId)
                 .SingleOrDefaultAsync();
 
+
```
Comment (Phil91) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalBackendDBAccess.cs:485]: Please remove the empty lines

File: src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/RegistrationController.cs [hunk 2]
```diff
@@ -27,12 +27,20 @@ public RegistrationController(IRegistrationBusinessLogic logic)
...
-    public Task<Pagination.Response<CompanyApplicationDetails>> GetApplicationDetailsAsync([FromQuery]int page, [FromQuery]int size) =>
+    public Task<Pagination.Response<CompanyApplicationDetails>> GetApplicationDetailsAsync([FromQuery] int page, [FromQuery] int size) =>
         _logic.GetCompanyApplicationDetailsAsync(page, size);
 
     [HttpPut]
     [Authorize(Roles = "approve_new_partner")]
     [Route("application/{applicationId}/approveRequest")]
     public Task<bool> ApprovePartnerRequest([FromRoute] Guid applicationId) =>
             _logic.ApprovePartnerRequest(applicationId);
+
```
Comment (Phil91) [inline src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/RegistrationController.cs:38]: Please add code documentation, for the swagger doc. You can see examples in this PR #91 or we can have a chat :)

**Review by Phil91 — CHANGES_REQUESTED** [review]

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalBackendDBAccess.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/PortalBackendDBAccess.cs]: fixed

File: src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/RegistrationController.cs (hunk 2 above)
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/RegistrationController.cs:38]: fixed

**Review by Phil91 — APPROVED** [review]

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 3]
```diff
@@ -116,4 +116,40 @@ await foreach (var user in _applicationRepository.GetWelcomeEmailDataUntrackedAs
...
+        var companyApplication = await _applicationRepository.GetCompanyAndApplicationForSubmittedApplication(applicationId).ConfigureAwait(false);
+        if (companyApplication == null)
+        {
+            throw new ArgumentException($"CompanyApplication {applicationId} is not in status SUBMITTED", "applicationId");
+        }
+        companyApplication.ApplicationStatusId = CompanyApplicationStatusId.DECLINED;
+        companyApplication.DateLastChanged = DateTimeOffset.UtcNow;
+        companyApplication.Company!.CompanyStatusId = CompanyStatusId.REJECTED;
+        await PostRegistrationCancelEmailAsync(applicationId).ConfigureAwait(false);
+        return true;
```
Comment (ntruchsess) [inline src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs:131]: the call to _portalRepositories.SaveAsync() is missing here.

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs [hunk 4]
```diff
@@ -103,4 +103,32 @@ public ApplicationRepository(PortalDbContext portalDbContext)
                         companyUser.Email,
                         companyUser.Company!.Name)))
             .AsAsyncEnumerable();
+
+    public async IAsyncEnumerable<WelcomeEmailData> GetRegistrationDeclineEmailDataUntrackedAsync(Guid applicationId, IDictionary<string, IEnumerable<string>> clientRoles)
+    {
+        foreach (var clientRole in clientRoles)
+        {
+            await foreach (var item in _dbContext.CompanyApplications
+                  .AsNoTracking()
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs:112]: indentation shall be 4 chars

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs [hunk 5]
```diff
@@ -103,4 +103,32 @@ public ApplicationRepository(PortalDbContext portalDbContext)
                         companyUser.Email,
                         companyUser.Company!.Name)))
             .AsAsyncEnumerable();
+
+    public async IAsyncEnumerable<WelcomeEmailData> GetRegistrationDeclineEmailDataUntrackedAsync(Guid applicationId, IDictionary<string, IEnumerable<string>> clientRoles)
+    {
+        foreach (var clientRole in clientRoles)
+        {
+            await foreach (var item in _dbContext.CompanyApplications
```
Comment (ntruchsess) [inline src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs:111]: if multiple roles are configured we may get duplicate entries in the result-set. I probably makes sense to first query the primary-keys of the roles in question (user_role_id) then pass all those roles at once to the query that selects the user data.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs (hunk 3 above)
Comment (qxz2mqe) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs:132]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs (hunk 4 above)
Comment (qxz2mqe) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs]: fixed

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs (hunk 5 above)
Comment (qxz2mqe) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/Repositories/ApplicationRepository.cs]: fixed

**Review by ntruchsess — APPROVED** [review]

**Comment by ntruchsess** [comment]
coding-wise the PR is fine. @qxz2mqe as we have a new configuration-parameter did you take care of updating the secrets accoridingly?

**Pull request merged**